openshift/dynamic/client.py

```diff
@@ -454,13 +386,225 @@ def __getattr__(self, name):
+        results = self.search(**kwargs)
+        if len(results) > 1 and kwargs.get('api_version'):
+            results = [
+                result for result in results if result.group_version == kwargs['api_version']
+            ]
+        if len(results) == 1:
+            return results[0]
+        elif not results:
+            raise ResourceNotFoundError('No matches found for {}'.format(kwargs))
+        else:
+            raise ResourceNotUniqueError('Multiple matches found for {}: {}'.format(kwargs, results))
+
+    def search(self, **kwargs):
+        return self.__search(self.__build_search(**kwargs), self.__resources, [])
+
+    def __search(self,  parts, resources, reqParams):
+        part = parts[0]
+        if part != '*':
+            resourcePart = resources.get(part)
+            if not resourcePart:
+                return []
+            elif isinstance(resourcePart, self.ResourceGroup):
+                if len(reqParams) != 2:
+                    raise ValueError("prefix and group params should be present, have %s" % reqParams)
+                # Check if we've requested resources for this group
+                if not resourcePart.resources:
+                    resourcePart.resources = self.get_resources_for_api_version(reqParams[0],
+                        reqParams[1], part, resourcePart.preferred)
+                return self.__search(parts[1:], resourcePart.resources, reqParams)
+            elif isinstance(resourcePart, dict):
+                # In this case parts [0] will be a specified prefix, group, version
+                # as we recurse
+                return self.__search(parts[1:], resourcePart, reqParams + [parts[0]] )
+            else:
+                if parts[1] != '*' and isinstance(parts[1], dict):
+                    for term, value in parts[1].items():
+                        if getattr(resourcePart, term) == value:
+                            return [resourcePart]
+                else:
+                    return [resourcePart]
+        else:
+            matches = []
+            for key in resources.keys():
+                matches.extend(self.__search([key] + parts[1:], resources, reqParams))
+            return matches
+
+    def __build_search(self, prefix=None, group=None, api_version=None, kind=None, **kwargs):
+        if not group and api_version and '/' in api_version:
+            group, api_version = api_version.split('/')
+
+        items = [prefix, group, api_version, kind, kwargs]
+        return list(map(lambda x: x or '*', items))
+
+
+    def __setup_resources(self):
+        groups = {}
+        groups['api'] = { '': {
+            'v1': self.ResourceGroup(True)
+        }}
+
+
+        if self.version.get('openshift'):
+            groups['oapi'] = { '': {
+                'v1': self.ResourceGroup(True)
+            }}
+
+        prefix = 'apis'
+        groups['apis'] = {}
+        groups_response = load_json(self.client.request('GET', '/{}'.format(prefix)))['groups']
```

Comment:
This is true, but one thing that gets a little trickier is handling the preferred field because AFAIK I can only get this in the groups response. That's why I went ahead and did it at the beginning because it's just one request. Any thoughts on this? Maybe the preferred field isn't needed? Or we could ignore it in some way until we have to do the groups request when either api_groups is accessed or get 404 on a request.

My original thinking was that since it's one request and in many cases we won't have all of the required fields to request the resource directly we might as well do it at the beginning, but more optimization is always nice.

Reply:
Yeah, I'm not sure how useful the preferred flag is tbh, I'm not sure it's really used. I would just ignore it for now.